// Copyright 2012 MaidSafe.net limited
//
// This MaidSafe Software is licensed under the MaidSafe.net Commercial License, version 1.0 or
// later, and The General Public License (GPL), version 3. By contributing code to this project You
// agree to the terms laid out in the MaidSafe Contributor Agreement, version 1.0, found in the root
// directory of this project at LICENSE, COPYING and CONTRIBUTOR respectively and also available at:
//
// http://www.novinet.com/license
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied. See the License for the specific language governing permissions and limitations under
// the License.

package maidsafe.lifestuff_manager.protobuf;

// All following messages are serialised into the payload field, and the message type added.
message WrapperMessage {
  required int32 type = 1;
  required bytes payload = 2;
  optional bytes message_signature = 3;
}

// LifeStuffManager receives this from a Client looking for the correct port and echos it back.
message ClientRegistrationRequest {
  required uint32 listening_port = 1;  // generated by ClientController
  required int32 version = 2;
}

message ClientRegistrationResponse {
  repeated bytes bootstrap_endpoint_ip = 1;
  repeated uint32 bootstrap_endpoint_port = 2;
  optional bytes path_to_new_installer = 3;
}

// LifeStuffManager receives this from Client and it will then start a vault.  The new vault will send
// a VaultIdentityRequest to the LifeStuffManager which will trigger the StartVaultResponse in reply to
// this message.
message StartVaultRequest {
  required bytes account_name = 1;
  required bytes pmid = 2;
  required bytes token = 3;
  required bytes token_signature = 4;
  required bool credential_change = 5;
  required uint32 client_port = 6;
  optional bytes chunkstore_path = 7;
  optional int32 identity_index = 8;
}

// LifeStuffManager sends this in response to a StartVaultRequest. It details whether or not the
// starting of the vault was successful.
message StartVaultResponse {
  required bool result = 1;
}

// LifeStuffManager receives this from Vault. Vault expects a VaultIdentityResponse.
message VaultIdentityRequest {
  required uint32 process_index = 1;  // as assigned by ProcessManager
  required uint32 listening_port = 2;  // generated by VaultController
  required int32 version = 3;
}

// LifeStuffManager sends this to a requesting Vault. It contains the ID that the Vault can assume,
// plus bootstrap nodes.
message VaultIdentityResponse {
  required bytes pmid = 1;
  required bytes chunkstore_path = 2;
  repeated bytes bootstrap_endpoint_ip = 3;
  repeated uint32 bootstrap_endpoint_port = 4;
}

// LifeStuffManager receives this from Vault once it has joined or failed to join the network and
// replies with the VaultJoinedNetworkAck.
message VaultJoinedNetwork {
  required uint32 process_index = 1;  // as assigned by ProcessManager
  required bool joined = 2;
}

message VaultJoinedNetworkAck {
  required bool ack = 1;
}

// Client receives this from LifeStuffManager once Vault has joined or failed to join the network and
// Client replies with the VaultJoinConfirmationAck.
message VaultJoinConfirmation {
  required bytes identity = 1;
  required bool joined = 2;
}

message VaultJoinConfirmationAck {
  required bool ack = 1;
}

//Vault sends this to LifeStuffManager when it discovers a suitable bootstrap node.
// LifeStuffManager replies with the SendEndpointToLifeStuffManagerResponse.
message SendEndpointToLifeStuffManagerRequest {
  required bytes bootstrap_endpoint_ip = 4;
  required uint32 bootstrap_endpoint_port = 5;
}

message SendEndpointToLifeStuffManagerResponse {
  required bool result = 1;
}

// LifeStuffManager receives this from Client and it will then stop vault with pmid.identity==identity
message StopVaultRequest {
  required bytes data = 1;
  required bytes signature = 2;
  required bytes identity = 3;
}

// LifeStuffManager sends this in response to a StopVaultRequest. It details whether or not the
// stopping of the vault was successful.
message StopVaultResponse {
  required bool result = 1;
}

// LifeStuffManager sends this to Vault
message VaultShutdownRequest {
  required uint32 process_index = 1;  // as assigned by ProcessManager
  required bytes data = 2;
  required bytes signature = 3;
}

// Vault replies with this to a requesting LifeStuffManager. It contains a boolean indicating whether
// or not the vault is shutting down.
message VaultShutdownResponse {
  required bool shutdown = 1;
  required uint32 process_index = 2;
}

// LifeStuffManager acknowledges VaultShutdownResponse with this.
message VaultShutdownResponseAck {
  required bool ack = 1;
}

// Client sends this to LifeStuffManager.  If the new_update_interval is set, the request is a setter,
// otherwise it's a getter.
message UpdateIntervalRequest {
  optional uint32 new_update_interval = 1;  // In seconds.
}

// LifeStuffManager sends this to a requesting Client indicating current update interval in seconds.
message UpdateIntervalResponse {
  required uint32 update_interval = 1;  // In seconds.
}

// LifeStuffManager sends this to Clients when a new version of the Client executable is available.
message NewVersionAvailable {
  required bytes new_version_filepath = 1;
}

// Client acknowledges NewVersionAvailable by replying with the same new_version_filepath that was
// passed in the NewVersionAvailable message.
message NewVersionAvailableAck {
  required bytes new_version_filepath = 1;
}

message BootstrapRequest {
  required uint32 message_id = 1;
}

message BootstrapResponse {
  repeated bytes bootstrap_endpoint_ip = 1;
  repeated uint32 bootstrap_endpoint_port = 2;
}
